Lib/hmac.py

```diff
@@ -50,31 +66,47 @@ def __init__(self, key, msg=None, digestmod=''):
         """
 
         if not isinstance(key, (bytes, bytearray)):
-            raise TypeError("key: expected bytes or bytearray, but got %r" % type(key).__name__)
+            raise TypeError(f"key: expected bytes or bytearray, "
+                            f"but got {type(key).__name__!r}")
 
         if not digestmod:
             raise TypeError("Missing required argument 'digestmod'.")
 
+        self.__init(key, msg, digestmod)
+
+    def __init(self, key, msg, digestmod):
         if _hashopenssl and isinstance(digestmod, (str, _functype)):
             try:
-                self._init_hmac(key, msg, digestmod)
+                self._init_openssl_hmac(key, msg, digestmod)
+                return
             except _hashopenssl.UnsupportedDigestmodError:
-                self._init_old(key, msg, digestmod)
-        else:
-            self._init_old(key, msg, digestmod)
+                pass
+        if _hmac and isinstance(digestmod, str):
```

Comment:
not needed initially for this feature (leave it for a followup), but we should probably do our best so support a _functype digestmod for this case as well by recognizing the specific hacl* built in hash function callables within _init_builtin_hmac and accepting those as well as the names.

Reply:
Yeah but since I think we need to find a way to share the logic for detecting this as it's currently only in hashlib. I tried in a first implementation but gave up afterwards :c

Reply:
See https://github.com/python/cpython/issues/131876.